File: leet-code-csharp/stack/MinStack.cs
namespace leet_code_csharp.stack;

public class MinStack
{
    private readonly Stack<int> _stack;
    private readonly Stack<int> _minStack;

    public MinStack()
    {
        _stack = new Stack<int>();
        _minStack = new Stack<int>();
    }

    public void Push( int val )
    {
        _stack.Push( val );

        int min = Math.Min( val, _minStack.Count == 0 ? val : _minStack.Peek() );
        _minStack.Push( min );
    }

    public void Pop()
    {
        _stack.Pop();
        _minStack.Pop();
    }

    public int Top()
    {
        return _stack.Pop();
    }

    public int GetMin()
    {
        return _minStack.Peek();
    }
}

File: cpp/leetcode-cpp/min-stack.cpp

using namespace std;

/*
    Design stack that supports push, pop, top, & retriving min element

    2 stacks, 1 normal & 1 monotonic decr, only push if lower than top

    Time: O(1)
    Space: O(n)
*/
class MinStack
{
    public:
    MinStack()
    {

    }

    void push( int val )
    {

    }

    void pop()
    {

    }

    int top()
    {

    }

    int getMin()
    {

    }
};

int main()
{
    MinStack* obj = new MinStack();
    obj->push( 10 );
    obj->pop();
    int param_3 = obj->top();
    int param_4 = obj->getMin();

    return 0;
}
